src/HW_models/NRF_DPPI.c

```diff
@@ -0,0 +1,224 @@
+ */
+void nrf_dppi_TASK_CHG_ENDIS( int groupnbr, bool enable /*false=disable task*/ )
+{
+  if ( enable )
+  {
+    bs_trace_raw_time(9, "dppi: Channel group %i enabled\n", groupnbr);
+    NRF_DPPI_regs.CHEN |= NRF_DPPI_regs.CHG[groupnbr];
+  } else
+  {
+    bs_trace_raw_time(9 ,"dppi: Channel group %i disable\n", groupnbr);
+    NRF_DPPI_regs.CHEN &= ~NRF_DPPI_regs.CHG[groupnbr];
+  }
+  //Note of the author: From the spec I cannot guess if these tasks will affect
+  //CHEN or a separate hidden register
+}
+
+void nrf_dppi_TASK_CHG0_EN() { nrf_dppi_TASK_CHG_ENDIS(0,true); }
+void nrf_dppi_TASK_CHG1_EN() { nrf_dppi_TASK_CHG_ENDIS(1,true); }
+void nrf_dppi_TASK_CHG2_EN() { nrf_dppi_TASK_CHG_ENDIS(2,true); }
+void nrf_dppi_TASK_CHG3_EN() { nrf_dppi_TASK_CHG_ENDIS(3,true); }
+void nrf_dppi_TASK_CHG4_EN() { nrf_dppi_TASK_CHG_ENDIS(4,true); }
+void nrf_dppi_TASK_CHG5_EN() { nrf_dppi_TASK_CHG_ENDIS(5,true); }
+void nrf_dppi_TASK_CHG0_DIS(){ nrf_dppi_TASK_CHG_ENDIS(0,false); }
+void nrf_dppi_TASK_CHG1_DIS(){ nrf_dppi_TASK_CHG_ENDIS(1,false); }
+void nrf_dppi_TASK_CHG2_DIS(){ nrf_dppi_TASK_CHG_ENDIS(2,false); }
+void nrf_dppi_TASK_CHG3_DIS(){ nrf_dppi_TASK_CHG_ENDIS(3,false); }
+void nrf_dppi_TASK_CHG4_DIS(){ nrf_dppi_TASK_CHG_ENDIS(4,false); }
+void nrf_dppi_TASK_CHG5_DIS(){ nrf_dppi_TASK_CHG_ENDIS(5,false); }
+
+/*Forward declaration*/
+struct subscr_list_elem;
+
+//Element of linked list of subscribers
+typedef struct subscr_list_elem
+{
+  dest_f_t task_function;
+  struct subscr_list_elem * p_next;
+} subscr_list_elem_t;
+
+//Table with TASKs subscribing to all channels
+static subscr_list_elem_t* dppi_subscriber_list[NUMBER_DPPI_CHANNELS];
+
+//Function for checking if an element is in the subscriber linked list
+static bool subscriber_list_check(dest_f_t task_function, subscr_list_elem_t ** pp_head)
+{
+  if (*pp_head == NULL)
+  {
+    return false;
+  }
+
+  subscr_list_elem_t *p_last_elem = *pp_head;
+  bool in_list = (p_last_elem->task_function == task_function);
+
+  while (p_last_elem->p_next != NULL)
+  {
+    if (p_last_elem->task_function == task_function)
+    {
+      in_list = true;
+    }
+    p_last_elem = p_last_elem->p_next;
+  }
+
+  return in_list;
+}
+
+//Function for appending an element to the subscriber linked list
+static void subscriber_list_append(subscr_list_elem_t ** pp_head, dest_f_t task_function)
+{
+  if (!subscriber_list_check(task_function, pp_head))
+  {
+    struct subscr_list_elem* p_elem = (struct subscr_list_elem*) malloc(sizeof(struct subscr_list_elem));
+    p_elem->task_function = task_function;
+    p_elem->p_next = (*pp_head);
+    *pp_head = p_elem;
+  }
+}
+
+//Function for removing an element from the subscriber linked list
+static void subscriber_list_remove(subscr_list_elem_t ** pp_head, dest_f_t task_function)
+{
+  if(subscriber_list_check(task_function, pp_head))
+  {
+    struct subscr_list_elem* temp = (struct subscr_list_elem*) malloc(sizeof(struct subscr_list_elem));
```

Comment:
You don't need this `malloc` and the `free`'s below. The temp variable is assigned in the line below, so the new memory is never used. There are some similar cases in functions below.